### PR TITLE
Proposal for pagination strategy to be applied to endpoints and provider...

### PR DESCRIPTION
**Pull request by vibe13**
...s

### CODE REVIEW

**Comment by vibe13** [comment]
@slaskawi @alexcreasy 
With this proposal, you would get i.e. all the products still using 
http://[host]:[port]/pnc-web/rest/product
while you could paginate and sort using i.e.
http://[host]:[port]/pnc-web/rest/product?pageIndex=1&pageSize=5&sorted_by=name&sorting=desc, getting a Pagination wrapper object with all the metadata of pagination taken from the Pageable object from Spring Data. These metadata in the wrapper object could be consumed on the client for the pagination behaviour.

Waiting for feedback about the implementation.

File: pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java [hunk 1]
```diff
@@ -24,8 +24,23 @@ public ProductProvider(ProductRepository productRepository) {
     public ProductProvider() {
     }
 
-    public List<ProductRest> getAll() {
-        return productRepository.findAll().stream().map(product -> new ProductRest(product)).collect(Collectors.toList());
+    @Override
+    public Function<? super Product, ? extends ProductRest> mapper() {
```
Comment (slaskawi) [inline pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java]: Maybe toRestModel would be a slightly better name

**Comment by slaskawi** [comment]
Looks great :)

File: pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java [hunk 2]
```diff
@@ -24,8 +24,23 @@ public ProductProvider(ProductRepository productRepository) {
...
+    public String getDefaultSortingField() {
+        return "name";
+    }
+
+    public Object getAll(Integer pageIndex, Integer pageSize, String field, String sorting) {
+
+        if (noPaginationRequired(pageIndex, pageSize, field, sorting)) {
+            return productRepository.findAll().stream().map(mapper()).collect(Collectors.toList());
+        } else {
+            return transform(productRepository.findAll(buildPageRequest(pageIndex, pageSize, field, sorting)));
```
Comment (slaskawi) [inline pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java]: Hmmm on the second thought, I see we are extracting everything from DB and apply pagination on those results. I think we should modify Datastore layer to accept `Pageable` objects. Just like it is done here: <http://docs.spring.io/spring-data/commons/docs/current/reference/html/#repositories.special-parameters>

File: pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java (hunk 1 above)
Comment (vibe13) [inline pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java]: :+1:

File: pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java (hunk 2 above)
Comment (vibe13) [inline pnc-rest/src/main/java/org/jboss/pnc/rest/provider/ProductProvider.java]: ok as clarified on IRC, a PageRequest is already fetched into the findAll() method, so Hibernate is doing the pagination work for us already.

**Pull request merged**